lib/charms/tempo_coordinator_k8s/v0/charm_tracing.py

```diff
@@ -398,45 +407,72 @@ def save(self, spans: typing.Sequence[ReadableSpan]):
         if self._max_event_history_length < 1:
             dev_logger.debug("buffer disabled: max history length < 1")
             return
-
-        current_history_length = len(self.load())
-        new_history_length = current_history_length + len(spans)
-        if (diff := self._max_event_history_length - new_history_length) < 0:
-            self.drop(diff)
         self._save(spans)
 
     def _serialize(self, spans: Sequence[ReadableSpan]) -> bytes:
         # encode because otherwise we can't json-dump them
         return encode_spans(spans).SerializeToString()
 
+    def _prune(self, queue: Sequence[bytes]):
+        """Prune the queue until it fits in our constraints."""
+        n_dropped_spans = 0
+        # drop older events if we are past the max history length
+        overflow = len(queue) - self._max_event_history_length
+        if overflow > 0:
+            n_dropped_spans += overflow
+            logger.error(
+                f"buffer exceeds max history length ({self._max_event_history_length} events)"
+            )
+
+        new_spans = deque(queue[-self._max_event_history_length :])
+
+        # drop older events if the buffer size in megabytes is too large
+        span_sizes = deque(len(span) for span in new_spans)
+        logged_drop = False
+        while sum(span_sizes) > self._max_buffer_size_mib * _MiB_TO_B:
+            span_sizes.popleft()
+            new_spans.popleft()
+            n_dropped_spans += 1
+            # only log this once
+            if logged_drop:
+                logger.error(f"buffer exceeds size limit ({self._max_buffer_size_mib}MiB)")
```

Comment:
```suggestion
                logger.warning(f"charm tracing buffer exceeds size limit ({self._max_buffer_size_mib}MiB). Older traces will be dropped.")
```

I'm thinking error might scare the charm operators as this doesn't mean  their charm is broken. Perhaps we don't need this log in other places than `dev_logger`?